audio/out/ao_sndio.c

```diff
@@ -0,0 +1,349 @@
+        goto error;
+    }
+    if (!sio_getpar(p->hdl, &p->par)) {
+        MP_ERR(ao, "couldn't get params\n");
+        goto error;
+    }
+    if (p->par.bps > 1 && p->par.le != SIO_LE_NATIVE) {
+        MP_ERR(ao, "swapped endian output not supported\n");
+        goto error;
+    }
+    if (p->par.bits == 8 && p->par.bps == 1 && !p->par.sig) {
+        ao->format = AF_FORMAT_U8;
+    } else if (p->par.bits == 16 && p->par.bps == 2 && p->par.sig) {
+        ao->format = AF_FORMAT_S16;
+    } else if ((p->par.bits == 32 || p->par.msb) && p->par.bps == 4 && p->par.sig) {
+        ao->format = AF_FORMAT_S32;
+    } else {
+        MP_ERR(ao, "couldn't set format\n");
+        goto error;
+    }
+
+    p->havevol = sio_onvol(p->hdl, volcb, p);
+    sio_onmove(p->hdl, movecb, p);
+
+    p->pfd = calloc (sio_nfds(p->hdl), sizeof (struct pollfd));
+    if (!p->pfd)
+        goto error;
+
+    ao->device_buffer = p->par.bufsz;
+
+    p->buffer = NULL;
+    p->playing = false;
+
+    return 0;
+
+error:
+    if (p->hdl)
+      sio_close(p->hdl);
+
+    return -1;
+}
+
+/*
+ * close device
+ */
+static void uninit(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    if (p->hdl)
+        sio_close(p->hdl);
+
+    free(p->pfd);
+}
+
+/*
+ * stop playing and empty buffers (for seeking/pause)
+ */
+static void reset(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    if (p->playing) {
+        MP_WARN(ao, "Blocking until remaining audio is played... (sndio design bug).\n");
+
+        p->playing = false;
+
+        if (!sio_stop(p->hdl))
+            MP_ERR(ao, "reset: couldn't stop\n");
+        p->delay = 0;
+    }
+}
+
+static void start(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+    int n;
+
+    p->playing = true;
+
+    if (!sio_start(p->hdl))
+        MP_ERR(ao, "start: couldn't start\n");
+
+    if (p->buffer) {
+        int buffer_old = p->buffer_pos;
+        while (p->buffer_pos > 0) {
+            n = sio_write(
+                    p->hdl,
+                    p->buffer + ao->sstride * (buffer_old - p->buffer_pos),
+                    p->buffer_pos * ao->sstride
+                ) / ao->sstride;
+
+            p->buffer_pos -= n;
+            p->delay += n;
+        }
+    }
+
+}
+
+static bool write(struct ao *ao, void **data, int samples)
+{
+    struct priv *p = ao->priv;
+    int n;
+
+    if (0 && !p->playing) {
+        sio_start(p->hdl);
+        p->playing = true;
+    }
+
+    if (p->playing) {
+        n = sio_write(p->hdl, data[0], samples * ao->sstride) / ao->sstride;
+        p->delay += n;
+    } else {
+        if (!p->buffer) {
+            p->buffer_pos = 0;
+            p->buffer = calloc(ao->sstride * p->par.bps , p->par.bufsz);
+            if (!p->buffer)
+                MP_ERR(ao, "couldn't allocate temporary buffer");
+        }
+
+        if (p->buffer) {
+            memcpy(p->buffer + p->buffer_pos, data[0], samples * ao->sstride);
+            p->buffer_pos += samples;
+        }
+
```

Comment:
In that code path probably better return ```true``` instead calling ```sio_eof(p->hdl)```.

Reply:
True